lib/Tools/circt-bmc/ExternalizeRegisters.cpp

```diff
@@ -0,0 +1,119 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/HW/HWInstanceGraph.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+#include "circt/Dialect/Verif/VerifOps.h"
+#include "circt/Support/Namespace.h"
+#include "circt/Tools/circt-bmc/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/FunctionCallUtils.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/PostOrderIterator.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+using namespace igraph;
+
+namespace circt {
+#define GEN_PASS_DEF_EXTERNALIZEREGISTERS
+#include "circt/Tools/circt-bmc/Passes.h.inc"
+} // namespace circt
+
+//===----------------------------------------------------------------------===//
+// Convert Lower To BMC pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct ExternalizeRegistersPass
+    : public circt::impl::ExternalizeRegistersBase<ExternalizeRegistersPass> {
+  using ExternalizeRegistersBase::ExternalizeRegistersBase;
+  void runOnOperation() override;
+};
+} // namespace
+
+void ExternalizeRegistersPass::runOnOperation() {
+  auto &instanceGraph = getAnalysis<hw::InstanceGraph>();
+  DenseSet<Operation *> handled;
+  DenseMap<StringAttr, SmallVector<Type>> addedInputs;
+  DenseMap<StringAttr, SmallVector<Type>> addedOutputs;
+
+  // Iterate over all instances in the instance graph. This ensures we visit
+  // every module, even private top modules (private and never instantiated).
+  for (auto *startNode : instanceGraph) {
+    if (handled.count(startNode->getModule().getOperation()))
+      continue;
+
+    // Visit the instance subhierarchy starting at the current module, in a
+    // depth-first manner. This allows us to inline child modules into parents
+    // before we attempt to inline parents into their parents.
+    for (InstanceGraphNode *node : llvm::post_order(startNode)) {
+      if (!handled.insert(node->getModule().getOperation()).second)
+        continue;
+
+      auto module =
+          dyn_cast_or_null<HWModuleOp>(node->getModule().getOperation());
+      if (!module)
+        continue;
+
+      mlir::Value firstClock;
+      unsigned numRegs = 0;
+      module->walk([&](Operation *op) {
+        if (auto regOp = dyn_cast<seq::CompRegOp>(op)) {
+          addedInputs[module.getSymNameAttr()].push_back(regOp.getType());
+          addedOutputs[module.getSymNameAttr()].push_back(
+              regOp.getInput().getType());
+          OpBuilder builder(regOp);
+          regOp.getResult().replaceAllUsesWith(
+              module.appendInput("", regOp.getType()).second);
```

Comment:
nit: could you add port names?

Reply:
It would be tricky to compute instance paths to each register so it's not necessary to do that within this PR but I believe we'll eventually want to see correspondence between ports and original registers for debuggability :smile:

Reply:
The names of the inputs & outputs are now based on the original names of the registers if present (suffixed with `_state` or `_input` depending on whether they represent the input to the register or its current state), otherwise it's just `reg_<n>_state`. Agreed instance paths would be nice in the future for debuggability!